File: TheSky/lua/op/LoginView/LoginServerView.lua
local _layer = nil
local t

LoginServerOwner = LoginServerOwner or {}
ccb["LoginServerOwner"] = LoginServerOwner

-- 最近登陆行数
local function lastLoginCellCount()
    return table.getTableCount(userdata.serverCodes) % 2 == 0 and math.floor(table.getTableCount(userdata.serverCodes) / 2) 
        or math.floor(table.getTableCount(userdata.serverCodes) / 2) + 1
end

-- 服务器行数
local function serverCellCount()
    return #userdata.serverListArr % 2 == 0 and math.floor(#userdata.serverListArr / 2) or math.floor(#userdata.serverListArr / 2) + 1
end

local function serverClick(tag)
    local server
    if tag >= 10000 then
        server = userdata.serverList[userdata.serverCodes[tostring(tag - 10000)]]
    else
        server = userdata.serverListArr[tag].v
    end
    userdata.selectServer = server
    userdata.serverCode = server.id
    getLoginLayer():showMain()
    getLoginMainLayer():menuEnabled(true)
end

local function setServerIcon( sprite,status )
    local frameName
    if status == 1 then
        frameName = "server_hot_icon.png"
    elseif status == 2 then
        frameName = "server_new_icon.png"
    elseif status == 3 then
        frameName = "server_full_icon.png"
    elseif status == 4 then
        frameName = "server_maintain_icon.png"
    elseif status == 5 then
        frameName = "server_recom_icon.png"
    end
    sprite = tolua.cast(sprite,"CCSprite")
    sprite:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName(frameName))
end

local function addServerTable()
    local serverLayer = tolua.cast(LoginServerOwner["serverLayer"], "CCLayer")
    -- @param fn string Callback type
    -- @param table LuaTableView
    -- @param a1 & a2 mixed Difference means for every "fn"
    local h = LuaEventHandler:create(function(fn, table, a1, a2)
        local r
        if fn == "cellSize" then
            r = CCSizeMake(serverLayer:getContentSize().width, 66)
        elseif fn == "cellAtIndex" then
            -- Return CCTableViewCell, a1 is cell index (zero based), a2 is dequeued cell (maybe nil)
            -- Do something to create cell and change the content
            
            if a2 then
                a2:removeAllChildrenWithCleanup(true)
            else 
                a2 = CCTableViewCell:create()
            end
            local offset = 0
            if userdata.serverCodes then
                offset = lastLoginCellCount() + 1
                if a1 == 0 then
                    local sp = CCSprite:createWithSpriteFrameName("lanchangtiao.png")
                    sp:setPosition(serverLayer:getContentSize().width / 2, 33)
                    a2:addChild(sp)
                    local text = CCSprite:createWithSpriteFrameName("zuijindenglu_text.png")
                    text:setPosition(serverLayer:getContentSize().width / 2, 33)
                    a2:addChild(text)
                elseif a1 <= lastLoginCellCount() then
                    local menu = CCMenu:create()
                    for i=0,1 do
                        local index = (a1 - 1) * 2 + i
                        local server = userdata.serverList[userdata.serverCodes[tostring(index)]]
                        if server then
                            local label = CCLabelTTF:create(server.serverName, "ccbResources/FZCuYuan-M03S.ttf", 30)
                            label:setColor(ccc3(184,45,18))
                            label:enableShadow(CCSizeMake(2,-2), 1, 0)
                            label:setHorizontalAlignment(kCCTextAlignmentLeft)
                            local hotSpr = CCSprite:createWithSpriteFrameName("server_hot_icon.png")
                            setServerIcon(hotSpr,server.status)

                            local norSp = CCSprite:createWithSpriteFrameName("server_btn_0.png")
                            local selSp = CCSprite:createWithSpriteFrameName("server_btn_1.png")
                            local item = CCMenuItemSprite:create(norSp, selSp)
                            item:addChild(label)
                            label:setAnchorPoint(ccp(0, 0.5))
                            label:setPosition(item:getContentSize().width * 0.29, item:getContentSize().height / 2)
                            item:setPosition(serverLayer:getContentSize().width / 4 * (i * 2 + 1), 30)
                            item:registerScriptTapHandler(serverClick)
                            menu:addChild(item, 0, index + 10000)
                            hotSpr:setPosition(hotSpr:getContentSize().width, item:getContentSize().height / 3 * 2)
                            item:addChild(hotSpr)
                        end
                    end
                    menu:setAnchorPoint(ccp(0, 0))
                    menu:setPosition(ccp(0, 0))
                    a2:addChild(menu)
                end
            end
            if a1 == offset then
                local sp = CCSprite:createWithSpriteFrameName("lanchangtiao.png")
                sp:setPosition(serverLayer:getContentSize().width / 2, 33)
                a2:addChild(sp)
                local text = CCSprite:createWithSpriteFrameName("suoyoufuwuqi_text.png")
                text:setPosition(serverLayer:getContentSize().width / 2, 33)
                a2:addChild(text)
            elseif a1 > offset and a1 <= offset + serverCellCount() then
                local menu = CCMenu:create()
                menu:setContentSize(CCSizeMake(serverLayer:getContentSize().width, 66))
                for i=1,2 do
                    local index = (a1 - offset - 1) * 2 + i
                    if userdata.serverListArr[index] then
                        local server = userdata.serverListArr[index].v
                        if server then
                            local label = CCLabelTTF:create(server.serverName, "ccbResources/FZCuYuan-M03S.ttf", 25)
                            label:setColor(ccc3(55,34,2))
                            label:enableShadow(CCSizeMake(2,-2), 1, 0)
                            label:setHorizontalAlignment(kCCTextAlignmentLeft)
                            local hotSpr = CCSprite:createWithSpriteFrameName("server_new_icon.png")
                            setServerIcon(hotSpr,server.status)

                            local norSp = CCSprite:createWithSpriteFrameName("server_btn_0.png")
                            local selSp = CCSprite:createWithSpriteFrameName("server_btn_1.png")
                            local item = CCMenuItemSprite:create(norSp, selSp)
                            item:addChild(label)
                            label:setAnchorPoint(ccp(0, 0.5))
                            label:setPosition(item:getContentSize().width * 0.29, item:getContentSize().height / 2)
                            item:setPosition(serverLayer:getContentSize().width / 4 * (i * 2 - 1), 30)
                            item:registerScriptTapHandler(serverClick)
                            menu:addChild(item, 0, index)
                            hotSpr:setPosition(hotSpr:getContentSize().width, item:getContentSize().height / 3 * 2)-- hotSpr:getContentSize().height / 2)
                            
                            item:addChild(hotSpr)
                        end
                    end
                end
                menu:setAnchorPoint(ccp(0, 0))
                menu:setPosition(ccp(0, 0))
                a2:addChild(menu)
            end
            r = a2
        elseif fn == "numberOfCells" then
            -- Return number of cells
            r = 0
            if userdata.serverCodes then
                r = lastLoginCellCount() + 1
            end 
            r = r + serverCellCount() + 1
        -- Cell events:
        elseif fn == "cellTouched" then         -- A cell was touched, a1 is cell that be touched. This is not necessary.
        elseif fn == "cellTouchBegan" then      -- A cell is touching, a1 is cell, a2 is CCTouch
        elseif fn == "cellTouchEnded" then      -- A cell was touched, a1 is cell, a2 is CCTouch
        elseif fn == "cellHighlight" then       -- A cell is highlighting, coco2d-x 2.1.3 or above
        elseif fn == "cellUnhighlight" then     -- A cell had been unhighlighted, coco2d-x 2.1.3 or above
        elseif fn == "cellWillRecycle" then     -- A cell will be recycled, coco2d-x 2.1.3 or above
        end
        return r
    end)
    
    local size = CCSizeMake(serverLayer:getContentSize().width, serverLayer:getContentSize().height * 0.98)
    t = LuaTableView:createWithHandler(h, size)
    t:setBounceable(true)
    t:setAnchorPoint(ccp(0,0))
    t:setPosition(ccp(0,0))
    t:setVerticalFillOrder(0)
    serverLayer:addChild(t)
    t:reloadData()
end



local function init()
    local cache = CCSpriteFrameCache:sharedSpriteFrameCache()
    cache:addSpriteFramesWithFile("ccbResources/login.plist")
    cache:addSpriteFramesWithFile("ccbResources/fontPic.plist")

    local  proxy = CCBProxy:create()
    local  node  = CCBReaderLoad("ccbResources/LoginServerView.ccbi",proxy,true,"LoginServerOwner")
    _layer = tolua.cast(node,"CCLayer")    

    local serverLayer = tolua.cast(LoginServerOwner["serverLayer"], "CCLayer")
    local grayBg = CCScale9Sprite:create("ccbResources/grayBg.png")
    grayBg:setContentSize(CCSizeMake(serverLayer:getContentSize().width * 0.98, serverLayer:getContentSize().height * 0.98))
    grayBg:setAnchorPoint(ccp(0.5, 0.5))
    grayBg:setPosition(ccp(serverLayer:getContentSize().width/2, serverLayer:getContentSize().height/2))
    serverLayer:addChild(grayBg, -1)
    addServerTable()

    local logo = tolua.cast(LoginServerOwner["logo"], "CCSprite")
    if isPlatform(IOS_APPLE2_ZH) 
        or isPlatform(IOS_KY_ZH) 
        or isPlatform(IOS_ITOOLS) 
        or isPlatform(IOS_PPZS_ZH) 
        or isPlatform(ANDROID_AGAME_ZH) 
        or isPlatform(IOS_91_ZH) 
        or isPlatform(ANDROID_91_ZH) 
        or isPlatform(ANDROID_360_ZH)
        or isPlatform(IOS_TBT_ZH)
        or isPlatform(ANDROID_COOLPAY_ZH)
        or isPlatform(ANDROID_GIONEE_ZH)
        or isPlatform(ANDROID_HTC_ZH) then

        if isSpringFestival then
            CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/springLogo.plist")
            logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("springLogo.png"))         
        else
            CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/tiantianLogo.plist")
            logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("tiantianLogo.png"))
        end

    elseif isPlatform(ANDROID_BAIDU_ZH) 
        or isPlatform(ANDROID_BAIDUIAPPPAY_ZH) 
        or isPlatform(ANDROID_MM_ZH) 
        or isPlatform(ANDROID_XIAOMI_ZH) 
        or isPlatform(ANDROID_OPPO_ZH) 
        or isPlatform(ANDROID_UC_ZH) then

        if isSpringFestival then
            CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/springLogo.plist")
            logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("springLogo.png"))         
        else
            CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/daLogo.plist")
            logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("daLogo.png"))
        end

    elseif isPlatform(ANDROID_JAGUAR_TC) then
        --todo
        CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/lbLogo.plist")
        --local logo = tolua.cast(LoginLoginOwner["logo"], "CCSprite")
        logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("lbLogo.png")) 
        
    elseif isPlatform(ANDROID_GV_MFACE_ZH) 
        or isPlatform(ANDROID_GV_MFACE_EN) 
        or isPlatform(ANDROID_GV_MFACE_EN_OUMEI)
        or isPlatform(ANDROID_GV_MFACE_EN_OUMEINEW)
        or isPlatform(IOS_GAMEVIEW_EN)
        or isPlatform(IOS_GVEN_BREAK) 
        or isPlatform(IOS_GAMEVIEW_ZH) then

        if isSpringFestival and (isPlatform(ANDROID_GV_MFACE_EN) or isPlatform(ANDROID_GV_MFACE_EN_OUMEI) or isPlatform(ANDROID_GV_MFACE_EN_OUMEINEW) or isPlatform(IOS_GAMEVIEW_EN) or isPlatform(IOS_GVEN_BREAK)) then
            CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/springLogo.plist")
            logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("gvSpringLogo.png"))         
        else
            CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/mfaceLogo.plist")
            logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("mfaceLogo.png"))
        end

    elseif isPlatform(IOS_MOBNAPPLE_EN) or isPlatform(ANDROID_VIETNAM_EN)
    or isPlatform(ANDROID_VIETNAM_EN_ALL) then
        if isSpringFestival then
            CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/springLogo.plist")
            logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("pfSpringLogo.png"))         
        else
            -- CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/login.plist")
            logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("logo_1.png"))
        end

    elseif isPlatform(ANDROID_GV_XJP_ZH)
        or isPlatform(ANDROID_GV_MFACE_TC)
        or isPlatform(IOS_GAMEVIEW_TC)
        or isPlatform(ANDROID_GV_MFACE_TC_GP) then

        CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/mfaceLogo2.plist")
        logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("mfaceLogo2.png"))

    elseif isPlatform(IOS_IIAPPLE_ZH) or isPlatform(IOS_KYPARK_ZH) or isPlatform(ANDROID_KY_ZH) or isPlatform(IOS_ITOOLSPARK)
        or isPlatform(IOS_PPZSPARK_ZH) or isPlatform(IOS_TBTPARK_ZH) or isPlatform(IOS_HAIMA_ZH) or isPlatform(IOS_AISIPARK_ZH) or isPlatform(IOS_XYZS_ZH)
        or isPlatform(ANDROID_XYZS_ZH) or isPlatform(IOS_DOWNJOYPARK_ZH) or isPlatform(ANDROID_DOWNJOY_ZH) then

        if isSpringFestival then
            CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/springLogo.plist")
            logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("springLogo.png"))         
        else
            CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/iappleLogo.plist")
            logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("iapple_logo.png"))
        end
    elseif isPlatform(IOS_TGAME_TC) or isPlatform(ANDROID_TGAME_TC) then
        if isSpringFestival then
            CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/springLogo.plist")
            logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("springLogo.png"))         
        else
            CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/tgTcLogo.plist")
            logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("tgTcLogo.png"))
        end
    elseif isPlatform(IOS_TGAME_TH) or isPlatform(ANDROID_TGAME_THAI) then
        if isSpringFestival then
            CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/springLogo.plist")
            logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("springLogo.png"))         
        else
            CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/tgthaiLogo.plist")
            logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("tgthaiLogo.png"))
        end
    else
        if isSpringFestival then
            CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/springLogo.plist")
            logo:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("springLogo.png"))         
        end
    end
end


function createLoginServerLayer()
    init()
    local function _onEnter()
    end

    local function _onExit()
        print("onExit")
        _layer = nil
        t = nil
    end

    --onEnter onExit
    local function layerEventHandler(eventType)
        if eventType == "enter" then
            if _onEnter then _onEnter() end
        elseif eventType == "exit" then
            if _onExit then _onExit() end
        end
    end
    -- addTestTable()
    _layer:registerScriptHandler(layerEventHandler)

    return _layer
end

File: TheSky/JavaScript/LZ/scenes/LoginScene/LoginServerView.js
var LoginServerView = cc.Layer.extend({
	ctor:function(){
		this._super();
		cc.BuilderReader.registerController("LoginServerOwner", {
		});
		this.node = cc.BuilderReader.load(ccbi_res.LoginServerView_ccbi, this);
		if(this.node != null) {
			this.addChild(this.node);
		}
		
		this.scList = LoginModule.getSelectedServerList();
		this.list = LoginModule.getServersList();
		this.selectedLen = Math.floor((this.scList.length + 1) / 2);
		this.listLen = Math.floor((this.list.length + 1) / 2);
		this.createTableView();
		var grayBg = new cc.Scale9Sprite(ccbi_dir + "/ccbi/ccbResources/grayBg.png");
		grayBg.setContentSize(cc.size(this.serverLayer.getContentSize().width * 0.98, this.serverLayer.getContentSize().height * 0.98));
		grayBg.setAnchorPoint(cc.p(0.5, 0.5));
		grayBg.setPosition(cc.p(this.serverLayer.getContentSize().width / 2, this.serverLayer.getContentSize().height / 2));
		this.serverLayer.addChild(grayBg, -1);
		
	},
	setServerIcon:function(sprite,status){
		var frameName;
		if (status === 1){
			frameName = "server_hot_icon.png"
		} else if (status === 2) {
			frameName = "server_new_icon.png"
		} else if (status === 3) {
			frameName = "server_full_icon.png"
		} else if (status === 4) {
			frameName = "server_maintain_icon.png"
		} else if (status === 5) {
			frameName = "server_recom_icon.png"
		}
	},
	serverClick:function(sender) {
		var tag = sender.getTag();
		var s;
		if (tag > 0 && tag <= this.scList.length === 0) {
			s = this.scList[tag - 1];
		} else if (tag > 0 && this.scList.length !== 0 && tag <= this.scList.length) {
			s = this.scList[tag - 1];
		} else if(tag > this.scList.length) {
			s = this.list[tag - 1 - this.scList.length];
		}
		if (!s) {
			return;
		}
		LoginModule.setSelectedServer(s.id);
	},
	createTableView:function() {
		if (this.tv) {
			var offset = this.tv.getContentOffset();
			this.tv.setContentOffset(offset, false);	
			return;
		}
		this.tv = new cc.TableView(this, this.serverLayer.getContentSize());
		this.tv.setDirection(cc.SCROLLVIEW_DIRECTION_VERTICAL);
		this.tv.setVerticalFillOrder(cc.TABLEVIEW_FILL_TOPDOWN);
		this.tv.attr({
			x:5,
			y:0,
			anchorX:0,
			anchorY:0
		});
		this.tv.setDelegate(this);
		this.serverLayer.addChild(this.tv);
	},
	scrollViewDisScroll:function(view){
	},
	scrollViewDidZoom:function(view){
	},
	tableCellTouched:function (table, cell) {
	},
	tableCellSizeForIndex:function (table, idx) {
		return cc.size(this.serverLayer.getContentSize().width, 80);
	},
	tableCellAtIndex:function (table, idx) {
		var cell = table.dequeueCell();
		if (!cell) {
			cell = new cc.TableViewCell();
		} else {
			cell.removeAllChildren(true);
		}
		if (idx === 0) {
			var bg = new cc.Sprite("#lanchangtiao.png");
			bg.setAnchorPoint(0,0);
			cell.addChild(bg);
			var text = new cc.Sprite("#zuijindenglu_text.png");
			text.setPosition(cc.p(bg.getContentSize().width/2 , bg.getContentSize().height / 2));
			text.setAnchorPoint(0.5, 0.5);
			bg.addChild(text);
		}else if (idx > 0 && idx <= this.selectedLen) {
			var menu = new cc.Menu();
			menu.setContentSize(cc.size(this.serverLayer.getContentSize().width, 80));
			cell.addChild(menu);
			menu.setAnchorPoint(0, 0);
			menu.setPosition(cc.p(0, 0));
			for (var i = 0; i < 2; i++) {
				var index = (idx - 1) * 2 + i;
				var serverIndex = index + 1;
				var server = this.scList[index];
				if (!server) {
					break;				
				}
				var item = new cc.MenuItemImage("#server_btn_0.png", "#server_btn_1.png", this.serverClick.bind(this), this);
				if (server.type === "1") {
					var hotSp = new cc.Sprite("#server_hot_icon.png");
					hotSp.setPosition(cc.p(hotSp.getContentSize().width, item.getContentSize().height / 3 * 2));
					item.addChild(hotSp);
				}
				item.setAnchorPoint(0, 0);
				item.setPosition(this.serverLayer.getContentSize().width / 2 * i, 0);
				menu.addChild(item, 0, serverIndex);
				
				var label = new cc.LabelTTF(server.serverName, FONT_NAME, 30);
				label.setColor(cc.color(184, 45,18));
				label.setAnchorPoint(0, 0.5);
				label.setPosition(cc.p(item.getContentSize().width * 0.29, item.getContentSize().height / 2));
				item.addChild(label);
			}
	
		} else if (idx === this.selectedLen + 1) {			
			var bg = new cc.Sprite("#lanchangtiao.png");
			bg.setAnchorPoint(0,0);
			cell.addChild(bg);
			var text = new cc.Sprite("#suoyoufuwuqi_text.png");
			text.setPosition(cc.p(bg.getContentSize().width/2 , bg.getContentSize().height / 2));
			text.setAnchorPoint(0.5, 0.5);
			bg.addChild(text);
		} else {
			var menu = new cc.Menu();
			menu.setContentSize(cc.size(cell.getContentSize().width, 66));
			for (var i = 0; i < 2; i++) {
				var index = (idx - 2 - this.selectedLen) * 2 + i;
				var serverIndex = this.scList.length + index + 1;
				var server = this.list[index];
				if (!server) {
					break;				
				}
				var norSp = new cc.Sprite("#server_btn_0.png");
				var selSp = new cc.Sprite("#server_btn_1.png");
				var item = new cc.MenuItemSprite(norSp, selSp, this.serverClick.bind(this), this);
				if (server.type === "1") {
					var hotSp = new cc.Sprite("#server_hot_icon.png");
					hotSp.setPosition(cc.p(hotSp.getContentSize().width, item.getContentSize().height / 3 * 2));
					item.addChild(hotSp);
				}
				item.setAnchorPoint(0, 0);
				item.setPosition(this.serverLayer.getContentSize().width / 2 * i, 0);
				menu.addChild(item, 0, serverIndex);
				
				var label = new cc.LabelTTF(server.serverName, FONT_NAME, 30);
				label.setColor(cc.color(184, 45,18));
				label.setAnchorPoint(0, 0.5);
				label.setPosition(cc.p(item.getContentSize().width * 0.29, item.getContentSize().height / 2));
				item.addChild(label);
			}	
			cell.addChild(menu);
			menu.setAnchorPoint(0, 0);
			menu.setPosition(cc.p(0, 0));
		}
		return cell;
	},
	numberOfCellsInTableView:function (table) {
		return this.selectedLen + this.listLen + 2;
	},
})